File: 217.contains-duplicate.cpp
/*
 * @lc app=leetcode id=217 lang=cpp
 *
 * [217] Contains Duplicate
 */

#include <bits/stdc++.h>

#include <cassert>

using namespace std;

// @lc code=start
class Solution {
 public:
  bool containsDuplicate(vector<int>& nums) {
    unordered_map<int, bool> map;
    for (auto num : nums) {
      if (map[num]) {
        return true;
      }
      map[num] = true;
    }
    return false;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> nums;
  nums = {1, 2, 3, 1};
  assert(s.containsDuplicate(nums) == true);
  nums = {1, 2, 3, 4};
  assert(s.containsDuplicate(nums) == false);
  nums = {1, 1, 1, 3, 3, 4, 3, 2, 4, 2};
  assert(s.containsDuplicate(nums) == true);
  return 0;
}

File: 217.contains-duplicate.rs
/*
 * @lc app=leetcode id=217 lang=rust
 *
 * [217] Contains Duplicate
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[macro_export]
macro_rules! list_node {
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let mut node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn contains_duplicate(nums: Vec<i32>) -> bool {
        let mut set = std::collections::HashSet::<i32>::new();
        for num in nums {
            if set.contains(&num) {
                return true;
            }
            set.insert(num);
        }
        false
    }
}
// @lc code=end

#[test]
fn test1() {
    assert!(Solution::contains_duplicate(vec![1, 2, 3, 1]));
}

#[test]
fn test2() {
    assert!(!Solution::contains_duplicate(vec![1, 2, 3, 4]));
}

#[test]
fn test3() {
    assert!(Solution::contains_duplicate(vec![
        1, 1, 1, 3, 3, 4, 3, 2, 4, 2
    ]),);
}
